crates/biome_css_analyze/src/lint/nursery/no_descending_specificity.rs

```diff
@@ -0,0 +1,209 @@
+    pub NoDescendingSpecificity {
+        version: "next",
+        name: "noDescendingSpecificity",
+        language: "css",
+        recommended: false,
+        sources: &[RuleSource::Stylelint("no-descending-specificity")],
+    }
+}
+
+#[derive(Debug)]
+pub struct DescendingSelector {
+    high: (TextRange, Specificity),
+    low: (TextRange, Specificity),
+}
+/// find tail selector
+/// ```css
+/// a b:hover {
+///   ^^^^^^^
+/// }
+/// ```
+fn find_tail_selector(selector: &AnyCssSelector) -> Option<String> {
+    match selector {
+        AnyCssSelector::CssCompoundSelector(s) => {
+            let simple = s
+                .simple_selector()
+                .map_or(String::new(), |s| s.syntax().text_trimmed().to_string());
+            let sub = s.sub_selectors().syntax().text_trimmed().to_string();
+
+            let last_selector = [simple, sub].join("");
+            Some(last_selector)
+        }
+        AnyCssSelector::CssComplexSelector(s) => {
+            s.right().as_ref().ok().and_then(find_tail_selector)
+        }
+        _ => None,
+    }
+}
+
+/// This function traverses the CSS rules starting from the given rule and checks for selectors that have the same tail selector.
+/// For each selector, it compares its specificity with the previously encountered specificity of the same tail selector.
+/// If a lower specificity selector is found after a higher specificity selector with the same tail selector, it records this as a descending selector.
+fn find_descending_selector(
+    rule: &CssSemanticRule,
+    model: &SemanticModel,
+    visited_rules: &mut BTreeSet<RuleId>,
+    visited_selectors: &mut BTreeMap<String, (TextRange, Specificity)>,
+    descending_selectors: &mut Vec<DescendingSelector>,
+) {
+    if visited_rules.contains(&rule.id) {
+        return;
+    } else {
+        visited_rules.insert(rule.id);
+    };
+
+    for selector in &rule.selectors {
+        let tail_selector = if let Some(s) = find_tail_selector(&selector.original) {
+            s
+        } else {
+            continue;
+        };
+
+        if let Some((last_textrange, last_specificity)) = visited_selectors.get(&tail_selector) {
+            if last_specificity > &selector.specificity {
+                descending_selectors.push(DescendingSelector {
+                    high: (*last_textrange, last_specificity.clone()),
+                    low: (selector.range, selector.specificity.clone()),
+                });
+            }
+        } else {
+            visited_selectors.insert(
+                tail_selector,
+                (selector.range, selector.specificity.clone()),
+            );
+        }
+    }
+
+    for child_rule in rule
+        .child_ids
+        .iter()
+        .filter_map(|id| model.get_rule_by_id(*id))
+    {
+        find_descending_selector(
+            child_rule,
+            model,
+            visited_rules,
+            visited_selectors,
+            descending_selectors,
+        );
+    }
```

Comment:
I think you can simply use match for better performance
```suggestion
    for child_rule in &rule.child_ids {
        if let Some(child_rule) = model.get_rule_by_id(*child_rule) {
            find_descending_selector(
                child_rule,
                model,
                visited_rules,
                visited_selectors,
                descending_selectors,
            );
        }
```

Reply:
Fixed at https://github.com/biomejs/biome/pull/4097/commits/0dbdbd1a03d8d48caf086edc83a4b7916bc5e7fe